x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/BlockFactory.java

```diff
@@ -0,0 +1,265 @@
+    public BooleanBlock.Builder newBooleanBlockBuilder(int estimatedSize) {
+        return new BooleanBlockBuilder(estimatedSize, this);
+    }
+
+    public BooleanBlock newBooleanArrayBlock(
+        boolean[] values,
+        int positionCount,
+        int[] firstValueIndexes,
+        BitSet nulls,
+        MvOrdering mvOrdering
+    ) {
+        var b = new BooleanArrayBlock(values, positionCount, firstValueIndexes, nulls, mvOrdering, this);
+        adjustBreaker(b.ramBytesUsed(), true);
+        return b;
+    }
+
+    public BooleanVector.Builder newBooleanVectorBuilder(int estimatedSize) {
+        return new BooleanVectorBuilder(estimatedSize, this);
+    }
+
+    public BooleanVector newBooleanArrayVector(boolean[] values, int positionCount) {
+        var b = new BooleanArrayVector(values, positionCount, this);
+        adjustBreaker(b.ramBytesUsed(), true);
+        return b;
+    }
+
+    public BooleanBlock newConstantBooleanBlockWith(boolean value, int positions) {
+        var b = new ConstantBooleanVector(value, positions, this).asBlock();
+        adjustBreaker(b.ramBytesUsed(), true);
+        return b;
+    }
+
+    public IntBlock.Builder newIntBlockBuilder(int estimatedSize) {
+        return new IntBlockBuilder(estimatedSize, this);
+    }
+
+    public IntBlock newIntArrayBlock(int[] values, int positionCount, int[] firstValueIndexes, BitSet nulls, MvOrdering mvOrdering) {
+        var b = new IntArrayBlock(values, positionCount, firstValueIndexes, nulls, mvOrdering, this);
+        adjustBreaker(b.ramBytesUsed(), true);
+        return b;
+    }
+
+    public IntVector.Builder newIntVectorBuilder(int estimatedSize) {
+        return new IntVectorBuilder(estimatedSize, this);
+    }
+
+    public IntVector newIntArrayVector(int[] values, int positionCount) {
+        var b = new IntArrayVector(values, positionCount, this);
+        adjustBreaker(b.ramBytesUsed(), true);
+        return b;
+    }
+
+    public IntBlock newConstantIntBlockWith(int value, int positions) {
+        var b = new ConstantIntVector(value, positions, this).asBlock();
+        adjustBreaker(b.ramBytesUsed(), true);
+        return b;
+    }
+
+    public LongBlock.Builder newLongBlockBuilder(int estimatedSize) {
+        return new LongBlockBuilder(estimatedSize, this);
+    }
+
+    public LongBlock newLongArrayBlock(long[] values, int positionCount, int[] firstValueIndexes, BitSet nulls, MvOrdering mvOrdering) {
+        var b = new LongArrayBlock(values, positionCount, firstValueIndexes, nulls, mvOrdering, this);
+        adjustBreaker(b.ramBytesUsed(), true);
+        return b;
+    }
+
+    public LongVector.Builder newLongVectorBuilder(int estimatedSize) {
+        return new LongVectorBuilder(estimatedSize, this);
+    }
+
+    public LongVector newLongArrayVector(long[] values, int positionCount) {
+        var b = new LongArrayVector(values, positionCount, this);
+        adjustBreaker(b.ramBytesUsed(), true);
+        return b;
+    }
```

Comment:
I imagine folks using this like:
```
blockFactory.adjustBreaker(positionCount * Long.BYTES, false);
long[] values = new long[positionCount];
for (int i = 0; i < positionCount; i++) {
  values[i] = doWhateverStuff
}
blockFactory.newLongArrayVectory(values, positionCount);
```

In a case like that we'd double count. Maybe in that case we'd want to call
```
blockFactory.newLongArrayVectory(values, positionCount, positionCount * Long.BYTES);
```
and let the builder take care of truing up the count with the reported `ramBytesUsed`.

This is probably fine to wait for a followup.

Reply:
yeah, I think that creating Vectors (not blocks) will be a particularly popular. The vector builders should be good enough and efficient for this kinda usage, but I do think that this usage is important to support - pre-adjusting the breaker, similar to what you outlined. 

Maybe we can make this a little more friendly. Here is an API sketch.  

```
long preAdjustedBytes = blockFactory.preAdjustBreakerForLong(positionCount);
long[] values = new long[positionCount];
for (int i = 0; i < positionCount; i++) {
  values[i] = doWhateverStuff
}
blockFactory.newLongArrayVector(values, positionCount, preAdjustedBytes);
```

Reply:
Something like that! I wonder if maybe we should wrap this kind of thing in a `LongVector.FixedLengthBuilder`. Basically the same as a `LongVector.Builder` but without the expanding. It's cook down to this but more meaningful names. Either way, that's a problem for another PR.